README.md
[![Build status][build-image]][build-url]
[![Tests coverage][cov-image]][cov-url]
[![npm version][npm-image]][npm-url]

# serverless-plugin-dynamodb-autoscaling

## Autoscaling configuration for DynamoDB tables

- **Convention over configuration approach** - Automatically discovers preconfigured tables and accompanies them with dedicated scaling resources. Configuration can be [fine tuned](#tables-configuration) addressing specific tables or indexes, or switched to [white list approach](#white-list-approach)
- **Resources savvy** - Existing project's IAM role is reused for handling scaling target resources. It's only mandatory `ScalableTarget` and `ScalingPolicy` resources that are added to CloudFormation stack

### Installation

```bash
npm install serverless-plugin-dynamodb-autoscaling
```

### Configuration

Activate plugin in `serverless.yml`

```yaml
plugins:
  - serverless-plugin-dynamodb-autoscaling
```

In most cases it's all you need to have autoscaling resources configured.
Still if you need to fine tune configuration for some reason see [tables](#tables-configuration) and [IAM role](#iam-role-configuration) configuration. Additionally if you approach any errors during CloudFormation deployment, refer to [troubleshooting](#troubleshooting)

#### Tables configuration

By default autoscaling configuration is automatically applied to all preconfigured DynamoDB tables and all its eventual global secondary indexes.

Still, we can exclude individual tables or tweak their configuration via configuration within`serverless.yml` config:

```yaml
resources:
  Resources:
    SomeTable1:
      Properties:
        TableName: foo
    SomeTable2:
      Properties:
        TableName: bar
    SomeTable3:
      Properties:
        TableName:
          Fn::Sub: ${AWS::Region}-test
custom:
  dynamodbAutoscaling:
    tablesConfig:
      # Disable autoscaling for table referenced by "SomeTable1" resource name
      SomeTable1: false

      # Disable autoscaling just for indexes of "SomeTable2" table
      SomeTable2:
        indexes: false

      # Tweak minCapacity setting for all tables of which resource names start with SomeTable
      # (glob patterns can be used)
      SomeTable*:
        minCapacity: 10

      SomeTable4:
        # Tweak maxCapacity setting for table referenced by "SomeTable4" (just table)
        table:
          maxCapacity: 300
        # Tweak targetUsage setting for SomeTable4 indexes
        indexes:
          targetUsage: 0.5

      SomeTable5:
        indexes:
          # Do not autoscale index 'foo'
          foo: false

      SomeTable6:
        indexes:
          # Do not autoscale any indexes but 'someIndex1' and 'someIndex2'
          "*": false
          someIndex1: true
          someIndex2:
            # Tweaking any of the configuration option will also whitelist the index
            minCapacity: 100
```

###### White list approach

If you prefer _white list_ instead of _black list_ approach you can handle configuration as below

```yaml
custom:
  dynamodbAutoscaling:
    tablesConfig:
      # Disable autoscaling for all
      "*": false

      # but enable for table referenced by resource name "SomeTable1"
      SomeTable1: true
```

##### Configurable settings:

- `maxCapacity` (defaults to `200`) refers to [`ScalableTarget.MaxCapacity`](https://docs.aws.amazon.com/autoscaling/application/APIReference/API_ScalableTarget.html)
- `minCapacity` (defaults to `5`) refers to [`ScalableTarget.MinCapacity`](https://docs.aws.amazon.com/autoscaling/application/APIReference/API_ScalableTarget.html)
- `targetUsage` (defaults to `0.75`) refers to [`ScalingPolicy.TargetTrackingScalingPolicyConfiguration.TargetValue`](https://docs.aws.amazon.com/autoscaling/application/APIReference/API_TargetTrackingScalingPolicyConfiguration.html) (value is multiplied by `100` when assigned to `TargetValue` setting)
- `scaleInCooldown` (defaults to `60`) refers to [`ScalingPolicy.TargetTrackingScalingPolicyConfiguration.ScaleInCooldown`](https://docs.aws.amazon.com/autoscaling/application/APIReference/API_TargetTrackingScalingPolicyConfiguration.html)
- `scaleOutCooldown` (defaults to `60`) refers to [`ScalingPolicy.TargetTrackingScalingPolicyConfiguration.ScaleOutCooldown`](https://docs.aws.amazon.com/autoscaling/application/APIReference/API_TargetTrackingScalingPolicyConfiguration.html)

#### `ScalingPolicy` chaining

By default `ScalingPolicy` resources are chained via `DependsOn` property, so they're deployed sequentially and not in parallel. It's to avoid reaching eventual CloudWatch rate limits.

Still it has a downside of slowing down the deployment. If number of tables in your stack is not large, or you've lifted rate limits on AWS side, then you can safely turn off that option to ensure more robust deployments:

```yaml
custom:
  dynamodbAutoscaling:
    chainScalingPolicies: false
```

#### IAM role configuration

By default existing lambda IAM role is reused (if recognized) or new dedicated IAM role is created.
Still it's possible to handle IAM configuration outside of this plugin, for that just pass ARN of externally configured IAM role via `iamRoleArn` setting:

```yaml
custom:
  dynamodbAutoscaling:
    iamRoleArn: "arn-of-iam-role-to-handle-tables"
```

### Troubleshooting

#### Eventual IAM policy update race condition issue

If at first deployment you're faced with one of the following errors:

- `Unable to assume IAM role`
- `Role is missing the following permissions`
- `The security token included in the request is invalid`

It can be result of a race condition issue described as following by AWS team:

_It's a known situation and confirmed by the internal team that manages CloudFormation that the propagation of IAM policies and resources might take longer than CloudFormation to launch the dependent resources. This race condition happens now and then, and unfortunately CloudFormation team is not able to determine programmatically when a role is effectively available in a region._

To workaround it, the stack with just IAM polices update (and no autoscaling resources yet) needs to be deployed first, and then further deployment may carry the autoscaling resources update (unfortunately just relying on `DependsOn` brings no rescue)

To make handling of that case easier this plugin enables the IAM only deployment via `iamOnly` option, you may refer to this option as one-time mean

```yaml
custom:
  dynamodbAutoscaling:
    iamOnly: true
```

### Tests

```bash
npm test
```

[build-image]: https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/workflows/Integrate/badge.svg
[build-url]: https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/actions?query=workflow%3AIntegrate
[cov-image]: https://img.shields.io/codecov/c/github/medikoo/serverless-plugin-dynamodb-autoscaling.svg
[cov-url]: https://codecov.io/gh/medikoo/serverless-plugin-dynamodb-autoscaling
[npm-image]: https://img.shields.io/npm/v/serverless-plugin-dynamodb-autoscaling.svg
[npm-url]: https://www.npmjs.com/package/serverless-plugin-dynamodb-autoscaling


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.1.1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v3.1.0...v3.1.1) (2023-05-10)

### Maintenance Improvements

- Confirm compatibitlity with versions v2 and v3 of the Framework ([644a952](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/644a95267136c5fec73c42c223b699a04d25550b))

<a name="3.1.0"></a>

# [3.1.0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v3.0.0...v3.1.0) (2019-01-22)

### Features

- make ScaleInCooldown and ScaleOutCooldown configurable ([fd3e365](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/fd3e365))

<a name="3.0.0"></a>

# [3.0.0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.2.0...v3.0.0) (2018-10-22)

### Features

- address tables by resource names in custom config ([286f1d5](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/286f1d5))

### BREAKING CHANGES

- Tables in configuration should now be addressed by CloudFormation resource name and not as it was so far table name (as passed in `TableName` property)

<a name="2.2.0"></a>

# [2.2.0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.1.1...v2.2.0) (2018-08-28)

### Features

- added iamRole option ([dfbdb84](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/dfbdb84))

<a name="2.1.1"></a>

## [2.1.1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.1.0...v2.1.1) (2018-02-08)

### Bug Fixes

- support serverless.yml with no custom block ([f412fd1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/f412fd1))

<a name="2.1.0"></a>

# [2.1.0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.0.3...v2.1.0) (2017-10-30)

### Features

- Allow to surpass ScalingPolicy chaining ([5812ca1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/5812ca1))

<a name="2.0.3"></a>

## [2.0.3](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.0.2...v2.0.3) (2017-10-23)

### Bug Fixes

- remove not necessary "autoscaling:\*" policy ([4e4a2ac](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/4e4a2ac))

<a name="2.0.2"></a>

## [2.0.2](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.0.1...v2.0.2) (2017-10-20)

### Bug Fixes

- ensure to grant "autoscaling:\*" rights ([f01ca10](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/f01ca10))

<a name="2.0.1"></a>

## [2.0.1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v2.0.0...v2.0.1) (2017-10-20)

### Bug Fixes

- generalize cloudwatch resources access ([7f8b302](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/7f8b302))

<a name="2.0.0"></a>

# [2.0.0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v1.0.1...v2.0.0) (2017-10-19)

### Bug Fixes

- narrow policy range ([7117722](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/7117722))
- remove "autoscaling:\*" policy from role ([5da01d0](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/5da01d0))

### Features

- allow non tables config related plugin settings ([5e29f67](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/5e29f67))
- apply policies unconditionally ([b0bffa7](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/b0bffa7))
- iamOnly option to address CF race condition ([7286716](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/7286716))
- improve IAM role handling ([193c55d](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/193c55d))
- reuse lambda role ([68ff842](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/68ff842))

### BREAKING CHANGES

- tables config was additionally nested with plugin settings at `tablesConfig` property.

Before:
custom:
dynamodbAutoscaling:
someTable: false

After:
custom:
dynamodbAutoscaling:
tablesConfig:
someTable: false

<a name="1.0.1"></a>

## [1.0.1](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/compare/v1.0.0...v1.0.1) (2017-10-16)

<a name="1.0.0"></a>

# 1.0.0 (2017-10-13)

### Bug Fixes

- CloudFormation no value handling ([16fa1ad](https://github.com/medikoo/serverless-plugin-dynamodb-autoscaling/commit/16fa1ad))


index.js
"use strict";

const compact       = require("es5-ext/array/#/compact")
    , assignDeep    = require("es5-ext/object/assign-deep")
    , copyDeep      = require("es5-ext/object/copy-deep")
    , isValue       = require("es5-ext/object/is-value")
    , isObject      = require("es5-ext/object/is-object")
    , objToArray    = require("es5-ext/object/to-array")
    , objForEach    = require("es5-ext/object/for-each")
    , objMap        = require("es5-ext/object/map")
    , capitalize    = require("es5-ext/string/#/capitalize")
    , hyphenToCamel = require("es5-ext/string/#/hyphen-to-camel")
    , d             = require("d")
    , lazy          = require("d/lazy")
    , minimatch     = require("minimatch");

const dimensionDefaults = {
	minCapacity: 5,
	maxCapacity: 200,
	targetUsage: 0.75,
	scaleInCooldown: 60,
	scaleOutCooldown: 60
};
const entityDefaults = { read: dimensionDefaults, write: dimensionDefaults }
    , indexesDefaults = { "*": entityDefaults }
    , tableDefaults = { table: entityDefaults, indexes: indexesDefaults };

const entitySettingNames = new Set(
	Object.keys(entityDefaults).concat(Object.keys(entityDefaults.read))
);

const resolveIndexToken = indexName =>
	indexName ? capitalize.call(hyphenToCamel.call(indexName).replace(/[^a-zA-Z0-9]/gu, "")) : "";

class ServerlessPluginDynamodbAutoscaling {
	constructor(serverless) {
		this.serverless = serverless;
		this.hooks = {
			"after:aws:package:finalize:mergeCustomProviderResources": this.configure.bind(this)
		};
	}
	configure() {
		if (!this.pluginConfig.iamRoleArn) {
			this.resources[this.iamRoleResourceId] = this.configureIamRole();
		}

		// iamOnly option offers workaround for race condition not handled as expected
		// by CloudFormation internals (more details on this case can be found in README.md)
		if (this.pluginConfig.iamOnly) return;

		Object.assign(this.resources, this.autoscalingResources);
	}

	// Configuration resolution
	resolveTableConfig(config) {
		if (!isValue(config)) return tableDefaults;
		if (!config) return false;
		if (!isObject(config)) return tableDefaults;
		if (!isValue(config.table) && !isValue(config.indexes)) {
			const entitySettings = this.resolveEntityConfig(config);
			return { table: entitySettings, indexes: { "*": entitySettings } };
		}
		return {
			table: this.resolveEntityConfig(config.table),
			indexes: this.resolveIndexesConfig(config.indexes)
		};
	}
	resolveIndexesConfig(config) {
		if (!isValue(config)) return indexesDefaults;
		if (!config) return false;
		if (!isObject(config)) return indexesDefaults;
		if (Object.keys(config).some(settingName => !entitySettingNames.has(settingName))) {
			const indexesSettings = { "*": this.resolveEntityConfig(config["*"]) };
			objForEach(config, (indexSettings, indexName) => {
				if (indexName === "*") return;
				indexesSettings[indexName] = this.resolveEntityConfig(indexSettings);
			});
			return indexesSettings;
		}
		return { "*": this.resolveEntityConfig(config) };
	}
	resolveEntityConfig(config) {
		if (!isValue(config)) return entityDefaults;
		if (!config) return false;
		if (!isObject(config)) return entityDefaults;
		if (!isValue(config.read) && !isValue(config.write)) {
			const dimensionSettings = this.resolveDimensionConfig(config);
			return { read: dimensionSettings, write: dimensionSettings };
		}
		return {
			read: this.resolveDimensionConfig(config.read),
			write: this.resolveDimensionConfig(config.write)
		};
	}
	resolveDimensionConfig(config) {
		if (!isValue(config)) return dimensionDefaults;
		if (!config) return false;
		if (!isObject(config)) return dimensionDefaults;
		return {
			minCapacity: config.minCapacity || dimensionDefaults.minCapacity,
			maxCapacity: config.maxCapacity || dimensionDefaults.maxCapacity,
			targetUsage: config.targetUsage || dimensionDefaults.targetUsage,
			scaleInCooldown: config.scaleInCooldown || dimensionDefaults.scaleInCooldown,
			scaleOutCooldown: config.scaleOutCooldown || dimensionDefaults.scaleOutCooldown
		};
	}

	// Autoscaling resources generation
	generateTableResources(resource, config) {
		const resources = {};
		if (config.table) {
			Object.assign(resources, this.generateEntityResources(config.table, resource.name));
		}
		if (!config.indexes) return resources;
		const globalSecondaryIndexes = resource.value.Properties.GlobalSecondaryIndexes;
		if (!Array.isArray(globalSecondaryIndexes)) return resources;
		for (const index of globalSecondaryIndexes) {
			const indexName = index.IndexName;
			Object.assign(
				resources,
				this.generateEntityResources(
					config.indexes[indexName] || config.indexes["*"], resource.name, indexName
				)
			);
		}
		return resources;
	}
	generateEntityResources(config, tableResourceName, indexName = null) {
		const resources = {};
		if (config.read) {
			Object.assign(
				resources,
				this.generateDimensionResources("read", config.read, tableResourceName, indexName)
			);
		}
		if (config.write) {
			Object.assign(
				resources,
				this.generateDimensionResources("write", config.write, tableResourceName, indexName)
			);
		}
		return resources;
	}
	generateDimensionResources(mode, config, tableResourceName, indexName = null) {
		const modeCapitalized = capitalize.call(mode);
		const resourceNamePrefix = `${ tableResourceName }${ resolveIndexToken(indexName) }${
			modeCapitalized
		}`;

		const targetResourceName = `${ resourceNamePrefix }ScalableTarget`;
		const policyResourceName = `${ resourceNamePrefix }ScalingPolicy`;
		const resourceAddress = ["table", { Ref: tableResourceName }];
		if (indexName) resourceAddress.push("index", indexName);

		const roleARN = this.pluginConfig.iamRoleArn || {
			"Fn::GetAtt": `${ this.iamRoleResourceId }.Arn`
		};

		const resources = {
			[targetResourceName]: {
				Type: "AWS::ApplicationAutoScaling::ScalableTarget",
				Properties: {
					MaxCapacity: config.maxCapacity,
					MinCapacity: config.minCapacity,
					ResourceId: { "Fn::Join": ["/", resourceAddress] },
					RoleARN: roleARN,
					ScalableDimension: `dynamodb:${ indexName ? "index" : "table" }:${
						modeCapitalized
					}CapacityUnits`,
					ServiceNamespace: "dynamodb"
				}
			},
			[policyResourceName]: {
				Type: "AWS::ApplicationAutoScaling::ScalingPolicy",
				DependsOn: this.lastPolicyResourceId,
				Properties: {
					PolicyName: `${ modeCapitalized }AutoScalingPolicy`,
					PolicyType: "TargetTrackingScaling",
					ScalingTargetId: { Ref: targetResourceName },
					TargetTrackingScalingPolicyConfiguration: {
						TargetValue: config.targetUsage * 100,
						ScaleInCooldown: config.scaleInCooldown,
						ScaleOutCooldown: config.scaleOutCooldown,
						PredefinedMetricSpecification: {
							PredefinedMetricType: `DynamoDB${ modeCapitalized }CapacityUtilization`
						}
					}
				}
			}
		};
		if (this.pluginConfig.chainScalingPolicies) this.lastPolicyResourceId = policyResourceName;
		return resources;
	}

	// IAM Role customization
	configureIamRoleResourceDynamodbAction() {
		const statements = this.iamRoleResource.Properties.Policies[0].PolicyDocument.Statement;
		let dynamodbStatement = statements.find(statement => {
			if (statement.Action.some(action => action === "dynamodb:*")) return true;
			return (
				statement.Action.some(action => action === "dynamodb:DescribeTable") &&
				statement.Action.some(action => action === "dynamodb:UpdateTable")
			);
		});
		if (!dynamodbStatement) {
			statements.push(
				(dynamodbStatement = {
					Effect: "Allow",
					Action: ["dynamodb:DescribeTable", "dynamodb:UpdateTable"],
					Resource: "arn:aws:dynamodb:*"
				})
			);
		}
	}
	configureIamRoleResourceCloudwatchAction() {
		const statements = this.iamRoleResource.Properties.Policies[0].PolicyDocument.Statement;
		let cloudwatchStatement = statements.find(statement =>
			statement.Action.some(action => action === "cloudwatch:*")
		);
		if (!cloudwatchStatement) {
			statements.push(
				(cloudwatchStatement = { Effect: "Allow", Action: ["cloudwatch:*"], Resource: "*" })
			);
		}
	}
	configureIamRole() {
		this.configureIamRolePolicyDocument();
		this.configureIamRolePolicies();
		return this.iamRoleResource;
	}
	configureIamRolePolicyDocument() {
		const services = this.iamRoleResource.Properties.AssumeRolePolicyDocument.Statement[0]
			.Principal.Service;
		if (services.includes("application-autoscaling.amazonaws.com")) return;
		services.push("application-autoscaling.amazonaws.com");
	}
	configureIamRolePolicies() {
		this.configureIamRoleResourceDynamodbAction();
		this.configureIamRoleResourceCloudwatchAction();
	}
}

Object.defineProperties(
	ServerlessPluginDynamodbAutoscaling.prototype,
	lazy({
		resources: d(function () {
			return this.serverless.service.provider.compiledCloudFormationTemplate.Resources;
		}),
		pluginConfig: d(function () {
			let pluginConfig = (this.serverless.service.custom || {}).dynamodbAutoscaling;
			if (!isValue(pluginConfig)) return { tablesConfig: {}, chainScalingPolicies: true };
			if (!isObject(pluginConfig)) {
				throw new Error(
					"Invalid 'dynamodbAutoscaling' configuration in serverless.yml. " +
						"Expected an object"
				);
			}
			pluginConfig = copyDeep(pluginConfig);
			if (!isValue(pluginConfig.tablesConfig)) {
				pluginConfig.tablesConfig = {};
			} else if (!isObject(pluginConfig.tablesConfig)) {
				throw new Error(
					"Invalid 'dynamodbAutoscaling.tablesConfig' configuration in serverless.yml. " +
						"Expected an object"
				);
			}
			if (!isValue(pluginConfig.chainScalingPolicies)) {
				pluginConfig.chainScalingPolicies = true;
			}
			return pluginConfig;
		}),
		tablesConfig: d(function () {
			const resolvedPluginConfig = objMap(this.pluginConfig.tablesConfig, config =>
				this.resolveTableConfig(config)
			);
			return Object.keys(this.resources)
				.map(resourceName => {
					const resource = this.resources[resourceName];
					if (resource.Type !== "AWS::DynamoDB::Table") return null;
					const configList = compact.call(
						objToArray(resolvedPluginConfig, (config, pattern) =>
							minimatch(resourceName, pattern) ? copyDeep(config) : null
						)
					);
					return {
						resource: { name: resourceName, value: resource },
						config: assignDeep({}, copyDeep(tableDefaults), ...configList)
					};
				})
				.filter(Boolean);
		}),
		autoscalingResources: d(function () {
			const autoscalingResources = {};
			for (const tableConfig of this.tablesConfig) {
				if (tableConfig.config) {
					Object.assign(
						autoscalingResources,
						this.generateTableResources(tableConfig.resource, tableConfig.config)
					);
				}
			}
			return autoscalingResources;
		}),
		iamRoleResourceId: d(function () {
			return this.resources.IamRoleLambdaExecution
				? "IamRoleLambdaExecution"
				: "DynamodbAutoscalingRole";
		}),
		iamRoleResource: d(function () {
			const defaultRole = {
				Type: "AWS::IAM::Role",
				Properties: {
					AssumeRolePolicyDocument: {
						Version: "2012-10-17",
						Statement: [
							{
								Effect: "Allow",
								Principal: { Service: [] },
								Action: ["sts:AssumeRole"]
							}
						]
					},
					Path: "/",
					Policies: [
						{
							PolicyName: "root",
							PolicyDocument: { Version: "2012-10-17", Statement: [] }
						}
					]
				}
			};
			return (this.resources[this.iamRoleResourceId] =
				this.resources.IamRoleLambdaExecution
					? this.resources[this.iamRoleResourceId]
					: assignDeep({}, defaultRole, this.resources[this.iamRoleResourceId] || {})
			);
		})
	})
);

module.exports = ServerlessPluginDynamodbAutoscaling;
